adr/ADR-42.md

```diff
@@ -0,0 +1,94 @@
+# NATS Client LifeCycle (-ERR Protocol Error Handling)
+
+| Metadata | Value          |
+| -------- | -------------- |
+| Date     | 2024-05-31     |
+| Author   | @aricart       |
+| Status   | Implemented    |
+| Tags     | client, server |
+
+## Context and Problem Statement
+
+Client lifecycle such as connect/reconnect/liveliness (ping/pong)/LDM behaviours
+are fairly complex in a NATS client. This ADR simply documents `-ERR` protocol
+messages that are sent to a client.
+
+The `-ERR` protocol message is an important signal for clients about things that
+are incorrect from the perspective of Permissions or Authorization.
+
+## Errors
+
+### Permission Violation
+
+`Permission Violation` means that the client tried to publish or subscribe on a
+subject for which it has no permissions. This type of error can happen or
+surface at any time, as changes to permissions intentionally or not can happen.
+
+The message will include `/(Publish|Subscription) to (\S+)/` this will indicate
+whether the error is related to a publish or subscription operation.
+
+For publish permission errors, it's hard to notify the client at the point of
+failure unless the client is synchronous. But the standard async error
+notification should be sufficient. In the case of request reply, since there's a
+subscription handling the response, this means that you can search subscriptions
+related to request and reply subjects, and notify them via the response
+mechanism for the request depending on the type of operation that was rejected.
+
+For subscription errors, a second level parse for `/using queue "(\S+)"/` will
+yield the queue if any that was used during the subscribe. This means that a
+client may have permissions on a subscription, but not in a specific queue or
+some other permutation.
+
+The server unfortunately doesn't make it easy for the client to know the actual
+subscription (SID) hosting the error but the logic for processing is simple:
+notify the first subscription that matches the subject and queue name (this
```

Comment:
Is it any different from notifying all subscriptions with matching subject and queue, closing them all and ignoring the subsequent errors (because they do not match any subscription in the client)? I'm not saying this is wrong, but genuinely curious if that would make any difference.

Reply:
The ones for the queue could be specific to the queue. As for it making a real difference not sure.